### PR TITLE
Backend Functionality of Times and Preferences Format

### PR DESCRIPTION
**Pull request by Domico-Carlo**
This pull request involves the following changes/additions:
- Implemented an algorithm to group students based on times and preferences data and upload this data to the database
- Modified database schema to be able to support the new data format
- Modified import process of timepref data in the database for easier fetching, and added support for more than 9 projects by refactoring the split(-1) used before
- Fixed issue of timestamps not importing correctly to the database
- Modified frontend for timepref group creation to not allow group sizes as it is not relevant

### CODE REVIEW

File: backend/helpers/analyticsRouteHandlerHelpers.js [hunk 1]
```diff
@@ -362,6 +514,45 @@ const getGroupAnalyticsEffort = async (unitCode, year, period, groupNumber) => {
...
+    // if (preferenceResults.length > 0) {
+    //     preferenceResults.forEach((result) => {
+    //          preferenceBarChartData["x"].push(result["preference"]);
+    //          preferenceBarChartData["y"].push(result["count"]);
+    //     });
+
+    //     preferenceAnalyticsData["data"].push(preferenceBarChartData);
+    // }
+    
+    return preferenceAnalyticsData
```
Comment (Rishi-Bidani) [inline backend/helpers/analyticsRouteHandlerHelpers.js:553]: Why are we returning preferenceBarChartData without filling out the x and y arrays? This function doesn't look complete?

File: backend/helpers/analyticsRouteHandlerHelpers.js [hunk 2]
```diff
@@ -362,6 +514,45 @@ const getGroupAnalyticsEffort = async (unitCode, year, period, groupNumber) => {
...
+        "personality title": "Preference Selection",
+        "description":
+            "Based on the input of each student, our algorithm attempts to a" +
+            "ssign each student their most preferred group. This is an aggre" +
+            "gation of how many students were assigned to their nth preferre" +
+            "d selection.",
+        data: [],
+    };
+
+    /** const preferenceResults = await promiseBasedQuery("FILL THIS SQL CODE IN THANKS" [unitCode, year, period, groupNumber]) */
```
Comment (Rishi-Bidani) [inline backend/helpers/analyticsRouteHandlerHelpers.js:533]: SQL not complete?

File: backend/routeHandlers/studentRouteHandler.js [hunk 3]
```diff
@@ -413,13 +413,205 @@ const addStudentEffort = async (personalityTestAttemptKeys, students) => {
...
+        // for each project 
+        for (let i = 0; i < preferencesArray.length; i++) {
+            resultInsertData.push([
+                preference_submission_id,
+                i,
+                preferencesArray[i]
+            ]);
+
+            // insert for each project 
+            // this looks like bad practice to insert in a for loop, im not sure how to fix since there will be N project_preference for N projects
```
Comment (Rishi-Bidani) [inline backend/routeHandlers/studentRouteHandler.js]: You can insert multiple rows in sql. If you build a string (or a list and make it a string at the end) and insert it after the for loop... but its finee for now
https://www.sqlservertutorial.net/sql-server-basics/sql-server-insert-multiple-rows/

File: backend/helpers/analyticsRouteHandlerHelpers.js (hunk 1 above)
Comment (Domico-Carlo) [reply backend/helpers/analyticsRouteHandlerHelpers.js:553]: This function was pulled from the analytics branch and should be replaced by the complete one that they have once they  do a pr as well

File: backend/helpers/analyticsRouteHandlerHelpers.js (hunk 2 above)
Comment (Domico-Carlo) [reply backend/helpers/analyticsRouteHandlerHelpers.js:533]: https://github.com/Monash-FIT3170/BaLance-Team-Forming-Dashboard/pull/147#discussion_r1716276152
Same issue

File: backend/routeHandlers/studentRouteHandler.js (hunk 3 above)
Comment (Domico-Carlo) [reply backend/routeHandlers/studentRouteHandler.js]: fixed i think

File: backend/helpers/analyticsRouteHandlerHelpers.js (hunk 1 above)
Comment (Domico-Carlo) [reply backend/helpers/analyticsRouteHandlerHelpers.js:553]: Problem handled by analytics branch.

**Review by Rishi-Bidani — APPROVED** [review]

**Pull request merged**